File: dsa_leetcode_beginner/0005-longest-palindromic-substring/0005-longest-palindromic-substring.java

class Solution {
    public String longestPalindrome(String s) {
        int n = s.length();
        if (n < 2) {
            return s;
        }
        boolean[][] dp = new boolean[n][n];
        
        // Initialize start position and max length of the palindrome
        int start = 0, maxLength = 1;

        // Base case: single character palindromes
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
        }

        // Base case: two-character palindromes
        for (int i = 0; i < n - 1; i++) {
            if (s.charAt(i) == s.charAt(i + 1)) {
                dp[i][i + 1] = true;
                start = i;
                maxLength = 2;
            }
        }

        // General case
        for (int length = 3; length <= n; length++) {
            for (int i = 0; i < n - length + 1; i++) {
                int j = i + length - 1;
                if (s.charAt(i) == s.charAt(j) && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    start = i;
                    maxLength = length;
                }
            }
        }

        return s.substring(start, start + maxLength);
    }
}

File: dsa_leetcode_beginner/0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n < 2:
            return s
        dp = [[False] * n for _ in range(n)]
        
        # Initialize start position and max length of the palindrome
        start, max_length = 0, 1

        # Base case: single character palindromes
        for i in range(n):
            dp[i][i] = True

        # Base case: two-character palindromes
        for i in range(n-1):
            if s[i] == s[i+1]:
                dp[i][i+1] = True                
            if dp[i][i+1]:
                start, max_length = i, 2

        # General case
        for length in range(3, n+1):  
            for i in range(n-length+1):
                j = i + length - 1
                if s[i] == s[j] and dp[i+1][j-1]:
                    dp[i][j] = True
                    start, max_length = i, length

        return s[start:start+max_length]
    
    
#Question: https://leetcode.com/problems/longest-palindromic-substring
#Blog: https://blog.unwiredlearning.com/longest-palindromic-substring